File: src/系统/09．表现系统/10．英雄语音/02．升级音效/01．英雄升级音效.lua
local ____lualib = require("lualib_bundle")
local __TS__StringTrim = ____lualib.__TS__StringTrim
local ____exports = {}
local ____00_FF0E_914D_7F6E = require("系统.09．表现系统.10．英雄语音.02．升级音效.00．配置")
local _____82F1_96C4_5347_7EA7_97F3_6548_914D_7F6E_5217_8868 = ____00_FF0E_914D_7F6E["英雄升级音效配置列表"]
---
-- @noSelfInFile
local jass = require("jass.common")
local ____require_result_0 = require("系统.00．核心系统.01．事件中心.06．英雄升级事件中心")
local registerHeroLevelListener = ____require_result_0.registerHeroLevelListener
local ____require_result_1 = require("lib.扩展函数.YDWE函数.09．YDUserData安全版")
local YDUserDataGetSafe = ____require_result_1.YDUserDataGetSafe
local ____require_result_2 = require("系统.01．单位系统.00．单位初始化创建.01．玩家英雄.01．玩家英雄配置工具")
local _____83B7_53D6_5355_4F4D_73A9_5BB6_82F1_96C4_914D_7F6E = ____require_result_2["获取单位玩家英雄配置"]
local ____require_result_3 = require("系统.01．单位系统.00．单位初始化创建.01．玩家英雄.03．玩家英雄别名")
local _____5355_4F4D_662F_5426_5339_914D_73A9_5BB6_82F1_96C4_540D_79F0 = ____require_result_3["单位是否匹配玩家英雄名称"]
local GetLocalPlayer = jass.GetLocalPlayer
local GetOwningPlayer = jass.GetOwningPlayer
local GetHandleId = jass.GetHandleId
local IsUnitType = jass.IsUnitType
local StartSound = jass.StartSound
local _____82F1_96C4_5347_7EA7_97F3_6548_5DF2_521D_59CB_5316 = false
local function _____53D6_5355_4F4D_5339_914D_540D_5217_8868(unit)
    local config = _____83B7_53D6_5355_4F4D_73A9_5BB6_82F1_96C4_914D_7F6E(unit)
    if config == nil then
        return {}
    end
    local result = {}
    local ____config_Name_4 = config.Name
    if ____config_Name_4 == nil then
        ____config_Name_4 = ""
    end
    local name = __TS__StringTrim(tostring(____config_Name_4))
    local ____config_Propernames_5 = config.Propernames
    if ____config_Propernames_5 == nil then
        ____config_Propernames_5 = ""
    end
    local proper = __TS__StringTrim(tostring(____config_Propernames_5))
    if name ~= "" then
        result[#result + 1] = name
    end
    if proper ~= "" then
        result[#result + 1] = proper
    end
    return result
end
local function _____5339_914D_5347_7EA7_97F3_6548_914D_7F6E(unit)
    do
        local i = 0
        while i < #_____82F1_96C4_5347_7EA7_97F3_6548_914D_7F6E_5217_8868 do
            local config = _____82F1_96C4_5347_7EA7_97F3_6548_914D_7F6E_5217_8868[i + 1]
            if _____5355_4F4D_662F_5426_5339_914D_73A9_5BB6_82F1_96C4_540D_79F0(unit, config["英雄名"]) then
                return config
            end
            i = i + 1
        end
    end
    return nil
end
local function _____672C_5730_64AD_653E_5347_7EA7_97F3_6548(unit, soundHandle)
    if soundHandle == nil then
        return
    end
    local owner = GetOwningPlayer(unit)
    if owner == nil or owner == 0 then
        return
    end
    if GetLocalPlayer() ~= owner then
        return
    end
    StartSound(soundHandle)
end
local function ____on_82F1_96C4_5347_7EA7_8BED_97F3(heroUnit)
    if heroUnit == nil or heroUnit == 0 then
        return
    end
    local owner = GetOwningPlayer(heroUnit)
    if owner == nil or owner == 0 then
        return
    end
    local registeredHero = YDUserDataGetSafe("player", owner, "英雄", "unit")
    if registeredHero == nil or registeredHero == 0 then
        return
    end
    if registeredHero ~= heroUnit and GetHandleId(registeredHero) ~= GetHandleId(heroUnit) then
        return
    end
    if IsUnitType(heroUnit, jass.UNIT_TYPE_DEAD) then
        return
    end
    local config = _____5339_914D_5347_7EA7_97F3_6548_914D_7F6E(heroUnit)
    if config == nil then
        return
    end
    _____672C_5730_64AD_653E_5347_7EA7_97F3_6548(heroUnit, config["播放音效"])
end
____exports["init英雄升级音效系统"] = function()
    if _____82F1_96C4_5347_7EA7_97F3_6548_5DF2_521D_59CB_5316 then
        return
    end
    _____82F1_96C4_5347_7EA7_97F3_6548_5DF2_521D_59CB_5316 = true
    registerHeroLevelListener(____on_82F1_96C4_5347_7EA7_8BED_97F3)
end
return ____exports

File: TS/系统/09．表现系统/10．英雄语音/02．升级音效/01．英雄升级音效.ts
/** @noSelfInFile */

const jass = require("jass.common") as any;

const { registerHeroLevelListener } = require("系统.00．核心系统.01．事件中心.06．英雄升级事件中心") as {
  registerHeroLevelListener: (this: void, callback: (this: void, heroUnit: any) => void) => void;
};
const { YDUserDataGetSafe } = require("lib.扩展函数.YDWE函数.09．YDUserData安全版") as {
  YDUserDataGetSafe: (this: void, tableType: string, tableKey: any, attr: string, valueType: string) => any;
};
const { 获取单位玩家英雄配置 } = require("系统.01．单位系统.00．单位初始化创建.01．玩家英雄.01．玩家英雄配置工具") as {
  获取单位玩家英雄配置: (this: void, unit: any) => Record<string, any> | null;
};
const { 单位是否匹配玩家英雄名称 } = require("系统.01．单位系统.00．单位初始化创建.01．玩家英雄.03．玩家英雄别名") as {
  单位是否匹配玩家英雄名称: (this: void, unit: any, name: string) => boolean;
};
import { 英雄升级音效配置列表, 英雄升级音效配置 } from "./00．配置";

const GetLocalPlayer = jass.GetLocalPlayer as (this: void) => any;
const GetOwningPlayer = jass.GetOwningPlayer as (this: void, unit: any) => any;
const GetHandleId = jass.GetHandleId as (this: void, handle: any) => number;
const IsUnitType = jass.IsUnitType as (this: void, unit: any, unitType: number) => boolean;
const StartSound = jass.StartSound as (this: void, soundHandle: any) => void;

let 英雄升级音效已初始化 = false;

function 取单位匹配名列表(this: void, unit: any): string[] {
  const config = 获取单位玩家英雄配置(unit);
  if (config == null) return [];
  const result: string[] = [];
  const name = String(config.Name ?? "").trim();
  const proper = String(config.Propernames ?? "").trim();
  if (name !== "") result.push(name);
  if (proper !== "") result.push(proper);
  return result;
}

function 匹配升级音效配置(this: void, unit: any): 英雄升级音效配置 | null {
  for (let i = 0; i < 英雄升级音效配置列表.length; i++) {
    const config = 英雄升级音效配置列表[i];
    if (单位是否匹配玩家英雄名称(unit, config.英雄名)) return config;
  }
  return null;
}

function 本地播放升级音效(this: void, unit: any, soundHandle: any): void {
  if (soundHandle == null) return;
  const owner = GetOwningPlayer(unit);
  if (owner == null || owner === 0) return;
  if (GetLocalPlayer() !== owner) return;
  StartSound(soundHandle);
}

function on英雄升级语音(this: void, heroUnit: any): void {
  if (heroUnit == null || heroUnit === 0) return;
  const owner = GetOwningPlayer(heroUnit);
  if (owner == null || owner === 0) return;
  const registeredHero = YDUserDataGetSafe("player", owner, "英雄", "unit");
  if (registeredHero == null || registeredHero === 0) return;
  if (registeredHero !== heroUnit && GetHandleId(registeredHero) !== GetHandleId(heroUnit)) return;
  if (IsUnitType(heroUnit, jass.UNIT_TYPE_DEAD)) return;
  const config = 匹配升级音效配置(heroUnit);
  if (config == null) return;
  本地播放升级音效(heroUnit, config.播放音效);
}

export function init英雄升级音效系统(this: void): void {
  if (英雄升级音效已初始化) return;
  英雄升级音效已初始化 = true;
  registerHeroLevelListener(on英雄升级语音);
}

export {};
